.travis.yml

```diff
@@ -5,7 +5,7 @@ jobs:
     - os: linux
       env: DEPLOY=true
       before_script:
-        - meson --unity=on --buildtype=release --strip -Ddebug=true --prefix=$PWD --bindir=$OTS-$TRAVIS_OS_NAME build
```

Comment:
We went through some length to make this work, and at least Firefox builds OTS in a semilar way, so I’d like to keep testing this on CI. Can you be more specific about the brotli build failure that prevent this?

Reply:
Sure. Here’s the output on my machine, with discussion below:

```sh
[ben@musicbox ots]$ bash -c 'export PKG_CONFIG_LIBDIR=/var/empty; meson build -Dunity=on && ninja -C build'
The Meson build system
Version: 0.55.3
Source dir: /home/ben/src/forks/ots
Build dir: /home/ben/src/forks/ots/build
Build type: native build
Project name: ots
Project version: 8.1.2
C compiler for the host machine: ccache cc (gcc 10.2.1 "cc (GCC) 10.2.1 20201125 (Red Hat 10.2.1-9)")
C linker for the host machine: cc ld.bfd 2.35-15
C++ compiler for the host machine: ccache c++ (gcc 10.2.1 "c++ (GCC) 10.2.1 20201125 (Red Hat 10.2.1-9)")
C++ linker for the host machine: c++ ld.bfd 2.35-15
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (1.7.3)
Found CMake: /usr/bin/cmake (3.18.4)
Run-time dependency freetype2 found: NO (tried pkgconfig and cmake)
Run-time dependency appleframeworks found: NO (tried framework)
Library gdi32 found: NO
Configuring config.h using configuration
Run-time dependency libbrotlidec found: NO (tried pkgconfig and cmake)
Looking for a fallback subproject for the dependency libbrotlidec

|Executing subproject google-brotli method meson 
|
|Project name: google-brotli
|Project version: 1.0.7
|C compiler for the host machine: ccache cc (gcc 10.2.1 "cc (GCC) 10.2.1 20201125 (Red Hat 10.2.1-9)")
|C linker for the host machine: cc ld.bfd 2.35-15
|Library m found: YES
|Checking for function "log2" with dependency -lm: YES 
|Build targets in project: 3
|Subproject google-brotli finished.

Dependency libbrotlidec from subproject subprojects/google-brotli found: YES 1.0.7
Run-time dependency zlib found: YES 1.2.11
Run-time dependency liblz4 found: NO (tried pkgconfig and cmake)
Looking for a fallback subproject for the dependency liblz4

|Executing subproject lz4 method meson 
|
|Project name: lz4
|Project version: 1.9.3
|C compiler for the host machine: ccache cc (gcc 10.2.1 "cc (GCC) 10.2.1 20201125 (Red Hat 10.2.1-9)")
|C linker for the host machine: cc ld.bfd 2.35-15
|Program GetLz4LibraryVersion.py found: YES
|Message: Project version is now: 1.9.3
|Compiler for C supports arguments -Wextra: YES 
|Compiler for C supports arguments -Wcast-qual: YES 
|Compiler for C supports arguments -Wcast-align: YES 
|Compiler for C supports arguments -Wshadow: YES 
|Compiler for C supports arguments -Wswitch-enum: YES 
|Compiler for C supports arguments -Wdeclaration-after-statement: YES 
|Compiler for C supports arguments -Wstrict-prototypes: YES 
|Compiler for C supports arguments -Wundef: YES 
|Compiler for C supports arguments -Wpointer-arith: YES 
|Compiler for C supports arguments -Wstrict-aliasing=1: YES 
|Build targets in project: 5
|Subproject lz4 finished.

Dependency liblz4 from subproject subprojects/lz4 found: YES 1.9.3
Has header "sys/time.h" : YES 
Run-time dependency GTest found: YES (prebuilt)
Program tests/test_bad_fonts.sh found: YES
Program tests/test_fuzzed_fonts.sh found: YES
Build targets in project: 11

ots 8.1.2

  Subprojects
    google-brotli: YES
              lz4: YES

Found ninja-1.10.1 at /usr/bin/ninja
ninja: Entering directory `build'                                                                   
[3/37] Compiling C object subprojects/brotli-1.0.7/libbrotli_encoder.a.p/meson-generated_brotli_encoder-unity1.c.o
FAILED: subprojects/brotli-1.0.7/libbrotli_encoder.a.p/meson-generated_brotli_encoder-unity1.c.o 
ccache cc -Isubprojects/brotli-1.0.7/libbrotli_encoder.a.p -Isubprojects/brotli-1.0.7 -I../subprojects/brotli-1.0.7 -I../subprojects/brotli-1.0.7/c/include -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -g -DOS_LINUX -fPIC -MD -MQ subprojects/brotli-1.0.7/libbrotli_encoder.a.p/meson-generated_brotli_encoder-unity1.c.o -MF subprojects/brotli-1.0.7/libbrotli_encoder.a.p/meson-generated_brotli_encoder-unity1.c.o.d -o subprojects/brotli-1.0.7/libbrotli_encoder.a.p/meson-generated_brotli_encoder-unity1.c.o -c subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:4:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment_two_pass.c:40:23: error: redefinition of ‘kHashMul32’
   40 | static const uint32_t kHashMul32 = 0x1E35A7BD;
      |                       ^~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:3:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment.c:41:23: note: previous definition of ‘kHashMul32’ was here
   41 | static const uint32_t kHashMul32 = 0x1E35A7BD;
      |                       ^~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:4:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment_two_pass.c:42:31: error: conflicting types for ‘Hash’
   42 | static BROTLI_INLINE uint32_t Hash(const uint8_t* p,
      |                               ^~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:3:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment.c:43:31: note: previous definition of ‘Hash’ was here
   43 | static BROTLI_INLINE uint32_t Hash(const uint8_t* p, size_t shift) {
      |                               ^~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:4:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment_two_pass.c:49:31: error: conflicting types for ‘HashBytesAtOffset’
   49 | static BROTLI_INLINE uint32_t HashBytesAtOffset(uint64_t v, size_t offset,
      |                               ^~~~~~~~~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:3:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment.c:48:31: note: previous definition of ‘HashBytesAtOffset’ was here
   48 | static BROTLI_INLINE uint32_t HashBytesAtOffset(
      |                               ^~~~~~~~~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:4:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment_two_pass.c:58:34: error: conflicting types for ‘IsMatch’
   58 | static BROTLI_INLINE BROTLI_BOOL IsMatch(const uint8_t* p1, const uint8_t* p2,
      |                                  ^~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:3:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment.c:58:34: note: previous definition of ‘IsMatch’ was here
   58 | static BROTLI_INLINE BROTLI_BOOL IsMatch(const uint8_t* p1, const uint8_t* p2) {
      |                                  ^~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:4:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment_two_pass.c:69:13: error: redefinition of ‘BuildAndStoreCommandPrefixCode’
   69 | static void BuildAndStoreCommandPrefixCode(
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:3:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment.c:127:13: note: previous definition of ‘BuildAndStoreCommandPrefixCode’ was here
  127 | static void BuildAndStoreCommandPrefixCode(const uint32_t histogram[128],
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:4:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment_two_pass.c:118:27: error: conflicting types for ‘EmitInsertLen’
  118 | static BROTLI_INLINE void EmitInsertLen(
      |                           ^~~~~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:3:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment.c:177:27: note: previous definition of ‘EmitInsertLen’ was here
  177 | static BROTLI_INLINE void EmitInsertLen(size_t insertlen,
      |                           ^~~~~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:4:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment_two_pass.c:148:27: error: conflicting types for ‘EmitCopyLen’
  148 | static BROTLI_INLINE void EmitCopyLen(size_t copylen, uint32_t** commands) {
      |                           ^~~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:3:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment.c:226:27: note: previous definition of ‘EmitCopyLen’ was here
  226 | static BROTLI_INLINE void EmitCopyLen(size_t copylen,
      |                           ^~~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:4:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment_two_pass.c:171:27: error: conflicting types for ‘EmitCopyLenLastDistance’
  171 | static BROTLI_INLINE void EmitCopyLenLastDistance(
      |                           ^~~~~~~~~~~~~~~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:3:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment.c:258:27: note: previous definition of ‘EmitCopyLenLastDistance’ was here
  258 | static BROTLI_INLINE void EmitCopyLenLastDistance(size_t copylen,
      |                           ^~~~~~~~~~~~~~~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:4:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment_two_pass.c:210:27: error: conflicting types for ‘EmitDistance’
  210 | static BROTLI_INLINE void EmitDistance(uint32_t distance, uint32_t** commands) {
      |                           ^~~~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:3:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment.c:301:27: note: previous definition of ‘EmitDistance’ was here
  301 | static BROTLI_INLINE void EmitDistance(size_t distance,
      |                           ^~~~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:4:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment_two_pass.c:222:13: error: redefinition of ‘BrotliStoreMetaBlockHeader’
  222 | static void BrotliStoreMetaBlockHeader(
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:3:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment.c:328:13: note: previous definition of ‘BrotliStoreMetaBlockHeader’ was here
  328 | static void BrotliStoreMetaBlockHeader(
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:4:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment_two_pass.c:521: warning: "MIN_RATIO" redefined
  521 | #define MIN_RATIO 0.98
      | 
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:3:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment.c:391: note: this is the location of the previous definition
  391 | #define MIN_RATIO 980
      | 
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:4:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment_two_pass.c:540:13: error: redefinition of ‘RewindBitPosition’
  540 | static void RewindBitPosition(const size_t new_storage_ix,
      |             ^~~~~~~~~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:3:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment.c:364:13: note: previous definition of ‘RewindBitPosition’ was here
  364 | static void RewindBitPosition(const size_t new_storage_ix,
      |             ^~~~~~~~~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:4:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment_two_pass.c:548:13: error: conflicting types for ‘EmitUncompressedMetaBlock’
  548 | static void EmitUncompressedMetaBlock(const uint8_t* input, size_t input_size,
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~
In file included from subprojects/brotli-1.0.7/libbrotli_encoder.a.p/brotli_encoder-unity1.c:3:
/home/ben/src/forks/ots/build/../subprojects/brotli-1.0.7/c/enc/compress_fragment.c:404:13: note: previous definition of ‘EmitUncompressedMetaBlock’ was here
  404 | static void EmitUncompressedMetaBlock(const uint8_t* begin, const uint8_t* end,
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~
[8/37] Compiling C object subprojects/brotli-1.0.7/libbrotli_encoder.a.p/meson-generated_brotli_encoder-unity0.c.o
ninja: build stopped: subcommand failed.
```

The problem is duplication in the encoder library between one-pass and two-pass implementations. This is kind of frustrating, since only the decoder is needed for ots. The brotli upstream has not been making many releases, so getting this changed there is not likely to happen quickly if at all.

I don’t know of any way to build only the dependencies that are actually used in a subproject. However, since the unity build is important, it’s probably possible to work around this by modifying the wrap file from the wrapdb, and providing a local version of the “patch” (using `patch_directory`, [introduced in meson 0.55](https://mesonbuild.com/Wrap-dependency-system-manual.html#accepted-configuration-properties-for-wraps) that builds only the decoder and comon libraries.

The same local wrap approach would be required for woff2 even once it is in the wrapdb, since the woff2 encoder would drag in the brotli encoder. The good news is that the local “fork” of each wrap would just be a meson.build file and a license file, which should be about as easy to maintain as the git submodule approach is today.

Reply:
Sounds good. We also don’t use brotli directly, so all this logic would be moved to woff2 build file.